src/clp_logging/readers.py

```diff
@@ -387,3 +388,251 @@ def __init__(
+                token_type: int
+                token_type, token, pos = CLPDecoder.decode_token(
+                    self.view[offset : self.valid_buf_len]
+                )
+                if token_type < 0:
+                    if token_type < -2:
+                        raise RuntimeError(
+                            f"Error decoding token: 0x{token.hex()}, type: {token_type}"
+                        )
+                    else:
+                        break  # Attempt to read more
+
+                # We have reached the end of stream.
+                if ID_EOF == token_type:
+                    break
+
+                log_buf += self.view[offset : offset + pos]
+                offset += pos
+
+                token_id: int = token_type & ID_MASK
+                if ID_TIMESTAMP == token_id:
+                    log_length: int = len(log_buf)
+                    if self.ostream_out_of_write_space(log_length):
+                        bytes_consumed: int = (
+                            self.total_bytes_read - log_length - (self.valid_buf_len - offset)
+                        )
+                        return bytes_consumed, self.generate_return_metadata()
+                    # Increment the last recorded timestamp
+                    self.last_timestamp_ms += int.from_bytes(token, BYTE_ORDER, signed=True)
+                    bytes_written = self.ostream.write(log_buf)
+                    if bytes_written <= 0:
+                        raise RuntimeError("Failed to write into out stream.")
+                    self.total_bytes_written += bytes_written
+                    log_buf.clear()
+
+            # Shift valid bytes to the start to make room for reading
+            # Grow the buffer if more than half is still valid
+            valid: int = self.valid_buf_len - offset
+            self.view[:valid] = self.view[offset : self.valid_buf_len]
+            if valid > len(self._buf) // 2:
+                tmp = bytearray(len(self._buf) * 2)
+                tmp[:valid] = self.view
+                self.view.release()
+                self._buf = tmp
+                self.view = memoryview(self._buf)
+
+            bytes_read: int = self.readinto_buf(valid)
+            if bytes_read == 0:
+                # No more to read. This is the end of the current segment.
+                self.ostream.write(EOF_CHAR)
+                # Log in the current buffer is not written into ostream.
+                bytes_consumed: int = self.total_bytes_read - len(log_buf) - valid
+                return bytes_consumed, self.generate_return_metadata()
+            elif bytes_read < 0:
+                raise RuntimeError("Failed to read from input stream.")
+
+            offset = 0
+            self.valid_buf_len = valid + bytes_read
+
+
+class CLPSegmentStreaming:
+    """
+    Wrapper for IR segment streaming. Provide encapsulation to the actual
+    stream reader class.
+    """
+
+    @staticmethod
+    def read(
```

Comment:
It is a bit unclear why `read` should be in a new class rather than a static method of `CLPSegmentStreamingReader`. Perhaps we could elaborating the doc string or additional unit tests would help make it clear.

Reply:
It encapsulates the actual reader class and ensure the read segment will not be called twice. The reader is not designed to be reuseable, all the class members are just the internal data structure used to maintain a single segment read operation. I've added this explanation in the code comments.